### PR TITLE
Fetch trading limits in base.exchange, removed fetch trading limits from has

### PR DESCRIPTION
**Pull request by samgermain**
- Moved `fetchTradingLimits` to base.exchange
- Removed `fetchTradingLimits` because it's not supposed to be called directly

----

```
% huobi fetchTradingLimits '["ETH/USDT"]'           
2022-01-06T05:40:59.281Z
Node.js: v14.17.0
CCXT v1.66.26
huobi.fetchTradingLimits (ETH/USDT)
280 ms
{
  'ETH/USDT': {
    info: {
      symbol: 'ethusdt',
      'buy-limit-must-less-than': '1.1',
      'sell-limit-must-greater-than': '0.9',
      'limit-order-must-greater-than': '0.001',
      'limit-order-must-less-than': '10000',
      'limit-buy-amount-must-less-than': '10000',
      'limit-sell-amount-must-less-than': '10000',
      'market-buy-order-must-greater-than': '5',
      'market-buy-order-must-less-than': '1000000',
      'market-sell-order-must-greater-than': '0.001',
      'market-sell-order-must-less-than': '1000',
      'limit-order-before-open-greater-than': '999999999',
      'limit-order-before-open-less-than': '0',
      'circuit-break-when-greater-than': '10000',
      'circuit-break-when-less-than': '10',
      'order-must-less-than': '0',
      'market-sell-order-rate-must-less-than': '0.1',
      'market-buy-order-rate-must-less-than': '0.1',
      'market-order-disabled-start-time': '0',
      'market-order-disabled-end-time': '0',
      'limit-order-limit-price-start-time': '0',
      'limit-order-limit-price-end-time': '0',
      'limit-order-limit-price-greater-than': '0.0',
      'limit-order-limit-price-less-than': '0.0',
      'equity-deviation-rate': null,
      'equity-deviation-rate-buy': null,
      'equity-deviation-rate-sell': null,
      'market-equity-deviation-rate': null,
      'market-equity-deviation-rate-buy': null,
      'market-equity-deviation-rate-sell': null,
      'equity-deviation-rate-switch': '0',
      'market-equity-deviation-rate-switch': '0',
      'limit-order-switch': '0',
      'limit-order-buy-offset': '0',
      'market-order-switch': '0',
      'market-order-buy-offset': '0'
    },
    limits: { amount: { min: 0.001, max: 10000 } }
  }
}
```

### CODE REVIEW

**Comment by samgermain** [comment]
On a side note, `eqonex.fetchTradingLimits` looks completely useless, and doesn't match the unified response

## equonex

```
% eqonex fetchTradingLimits '["ETH/USDT"]'
2022-01-06T05:46:56.146Z
Node.js: v14.17.0
CCXT v1.66.26
eqonex.fetchTradingLimits (ETH/USDT)
336 ms
{
  info: [],
  limits: {
    amount: { min: undefined, max: undefined },
    price: { min: undefined, max: undefined },
    cost: { min: undefined, max: undefined }
  }
}
```

---

```
module.exports = class eqonex extends Exchange {
...
    async fetchTradingLimits (symbols = undefined, params = {}) {
        await this.loadMarkets ();
        // getExchangeInfo
        const response = await this.publicGetGetExchangeInfo (params);
        const tradingLimits = this.safeValue (response, 'tradingLimits', []);
        // To-do parsing response when available
        return {
            'info': tradingLimits,
            'limits': {
                'amount': {
                    'min': undefined,
                    'max': undefined,
                },
                'price': {
                    'min': undefined,
                    'max': undefined,
                },
                'cost': {
                    'min': undefined,
                    'max': undefined,
                },
            },
        };
    }
    ...
}
 ```
 
 ## cdax
 
 ```
 % cdax fetchTradingLimits '["ETH/USDT"]'
2022-01-06T05:47:14.460Z
Node.js: v14.17.0
CCXT v1.66.26
cdax.fetchTradingLimits (ETH/USDT)
278 ms
{
  'ETH/USDT': {
    info: {
      symbol: 'ethusdt',
      'buy-limit-must-less-than': '1.1',
      'sell-limit-must-greater-than': '0.9',
      'limit-order-must-greater-than': '0.001',
      'limit-order-must-less-than': '10000',
      'limit-buy-amount-must-less-than': '10000',
      'limit-sell-amount-must-less-than': '10000',
      'market-buy-order-must-greater-than': '5',
      'market-buy-order-must-less-than': '1000000',
      'market-sell-order-must-greater-than': '0.001',
      'market-sell-order-must-less-than': '1000',
      'limit-order-before-open-greater-than': '999999999',
      'limit-order-before-open-less-than': '0',
      'circuit-break-when-greater-than': '10000',
      'circuit-break-when-less-than': '10',
      'order-must-less-than': '0',
      'market-sell-order-rate-must-less-than': '0.1',
      'market-buy-order-rate-must-less-than': '0.1',
      'market-order-disabled-start-time': '0',
      'market-order-disabled-end-time': '0',
      'limit-order-limit-price-start-time': '0',
      'limit-order-limit-price-end-time': '0',
      'limit-order-limit-price-greater-than': '0.0',
      'limit-order-limit-price-less-than': '0.0',
      'equity-deviation-rate': null,
      'equity-deviation-rate-buy': null,
      'equity-deviation-rate-sell': null,
      'market-equity-deviation-rate': null,
      'market-equity-deviation-rate-buy': null,
      'market-equity-deviation-rate-sell': null,
      'equity-deviation-rate-switch': '0',
      'market-equity-deviation-rate-switch': '0',
      'limit-order-switch': '0',
      'limit-order-buy-offset': '0',
      'market-order-switch': '0',
      'market-order-buy-offset': '0'
    },
    limits: { amount: { min: 0.001, max: 10000 } }
  }
}
```

---

## Huobi

```
% huobi fetchTradingLimits '["ETH/USDT"]'
2022-01-06T05:46:47.191Z
Node.js: v14.17.0
CCXT v1.66.26
huobi.fetchTradingLimits (ETH/USDT)
256 ms
{
  'ETH/USDT': {
    info: {
      symbol: 'ethusdt',
      'buy-limit-must-less-than': '1.1',
      'sell-limit-must-greater-than': '0.9',
      'limit-order-must-greater-than': '0.001',
      'limit-order-must-less-than': '10000',
      'limit-buy-amount-must-less-than': '10000',
      'limit-sell-amount-must-less-than': '10000',
      'market-buy-order-must-greater-than': '5',
      'market-buy-order-must-less-than': '1000000',
      'market-sell-order-must-greater-than': '0.001',
      'market-sell-order-must-less-than': '1000',
      'limit-order-before-open-greater-than': '999999999',
      'limit-order-before-open-less-than': '0',
      'circuit-break-when-greater-than': '10000',
      'circuit-break-when-less-than': '10',
      'order-must-less-than': '0',
      'market-sell-order-rate-must-less-than': '0.1',
      'market-buy-order-rate-must-less-than': '0.1',
      'market-order-disabled-start-time': '0',
      'market-order-disabled-end-time': '0',
      'limit-order-limit-price-start-time': '0',
      'limit-order-limit-price-end-time': '0',
      'limit-order-limit-price-greater-than': '0.0',
      'limit-order-limit-price-less-than': '0.0',
      'equity-deviation-rate': null,
      'equity-deviation-rate-buy': null,
      'equity-deviation-rate-sell': null,
      'market-equity-deviation-rate': null,
      'market-equity-deviation-rate-buy': null,
      'market-equity-deviation-rate-sell': null,
      'equity-deviation-rate-switch': '0',
      'market-equity-deviation-rate-switch': '0',
      'limit-order-switch': '0',
      'limit-order-buy-offset': '0',
      'market-order-switch': '0',
      'market-order-buy-offset': '0'
    },
    limits: { amount: { min: 0.001, max: 10000 } }
  }
}
```

**Comment by samgermain** [comment]
@frosty00 @kroitor 

Let me know if it's worth implementing `fetchTradingLimitsById` on binance

**Comment by kroitor** [comment]
Plz, check this exchange: https://github.com/ccxt/ccxt/blob/master/js/eqonex.js#L1208 (it's an inconsistency, we will have to do something about it).

**Comment by kroitor** [comment]
@samgermain 

> Let me know if it's worth implementing `fetchTradingLimitsById` on binance

If you can – please, go ahead.

**Comment by samgermain** [comment]
> Plz, check this exchange: https://github.com/ccxt/ccxt/blob/master/js/eqonex.js#L1208 (it's a pre-existing inconsistency, we will have to do something about it).

https://github.com/ccxt/ccxt/pull/11140#issuecomment-1006303113

File: python/ccxt/base/exchange.py [hunk 1]
```diff
@@ -2638,3 +2638,18 @@ def handle_market_type_and_params(self, method_name, market=None, params={}):
         type = self.safe_string_2(params, 'defaultType', 'type', market_type)
         params = self.omit(params, ['defaultType', 'type'])
         return [type, params]
+
+    def fetchTradingLimits(self, symbols=None, params={}):
```
Comment (kroitor) [inline python/ccxt/base/exchange.py:2642]: This has to be overrided for coroutines in the async version in ccxt/async_support/base/exchange.py.

Comment (samgermain) [reply python/ccxt/base/exchange.py:2648]: how do we test the async methods?

Comment (kroitor) [reply python/ccxt/base/exchange.py:2648]: 1. If the unified method is a part of the existing testset:
```bash
node js/test/test exchange_id
python3 python/ccxt/test/test_async.py exchange_id
php -f php/test/test_async.php exchange_id
```
2. If we don't have a corresponding test yet – we should make it, or we should make an ad-hoc script for the testing purpose, in-place.

**Comment by samgermain** [comment]
You'll want to double check this, I'm not sure if everything is done right, in `parseTradingLimits` I overwrite `symbol` for example

```
% binance fetchTradingLimitsById BTCUSDT --details
2022-01-08T03:31:48.474Z
Node.js: v14.17.0
CCXT v1.66.42
binance.fetchTradingLimitsById (BTCUSDT)
633 ms
{
  'BTC/USDT': {
    info: {
      symbol: 'BTCUSDT',
      status: 'TRADING',
      baseAsset: 'BTC',
      baseAssetPrecision: '8',
      quoteAsset: 'USDT',
      quotePrecision: '8',
      quoteAssetPrecision: '8',
      baseCommissionPrecision: '8',
      quoteCommissionPrecision: '8',
      orderTypes: [
        'LIMIT',
        'LIMIT_MAKER',
        'MARKET',
        'STOP_LOSS_LIMIT',
        'TAKE_PROFIT_LIMIT'
      ],
      icebergAllowed: true,
      ocoAllowed: true,
      quoteOrderQtyMarketAllowed: true,
      isSpotTradingAllowed: true,
      isMarginTradingAllowed: true,
      filters: [
        {
          filterType: 'PRICE_FILTER',
          minPrice: '0.01000000',
          maxPrice: '1000000.00000000',
          tickSize: '0.01000000'
        },
        {
          filterType: 'PERCENT_PRICE',
          multiplierUp: '5',
          multiplierDown: '0.2',
          avgPriceMins: '5'
        },
        {
          filterType: 'LOT_SIZE',
          minQty: '0.00001000',
          maxQty: '9000.00000000',
          stepSize: '0.00001000'
        },
        {
          filterType: 'MIN_NOTIONAL',
          minNotional: '10.00000000',
          applyToMarket: true,
          avgPriceMins: '5'
        },
        { filterType: 'ICEBERG_PARTS', limit: '10' },
        {
          filterType: 'MARKET_LOT_SIZE',
          minQty: '0.00000000',
          maxQty: '186.34181397',
          stepSize: '0.00000000'
        },
        { filterType: 'MAX_NUM_ORDERS', maxNumOrders: '200' },
        { filterType: 'MAX_NUM_ALGO_ORDERS', maxNumAlgoOrders: '5' }
      ],
      permissions: [ 'SPOT', 'MARGIN' ]
    },
    limits: {
      price: { min: 0.01, max: 1000000 },
      amount: { min: 0.00001, max: 9000 },
      market: { min: 0, max: 186.34181397 },
      cost: { min: 10 }
    }
  }
}
```

**Comment by samgermain** [comment]
> @samgermain
> 
> > Let me know if it's worth implementing `fetchTradingLimitsById` on binance
> 
> If you can – please, go ahead.

You'll want to double check this, I'm not sure if everything is done right, in `parseTradingLimits` I overwrite `symbol` for example

```
% binance fetchTradingLimitsById BTCUSDT --details
2022-01-08T03:31:48.474Z
Node.js: v14.17.0
CCXT v1.66.42
binance.fetchTradingLimitsById (BTCUSDT)
633 ms
{
  'BTC/USDT': {
    info: {
      symbol: 'BTCUSDT',
      status: 'TRADING',
      baseAsset: 'BTC',
      baseAssetPrecision: '8',
      quoteAsset: 'USDT',
      quotePrecision: '8',
      quoteAssetPrecision: '8',
      baseCommissionPrecision: '8',
      quoteCommissionPrecision: '8',
      orderTypes: [
        'LIMIT',
        'LIMIT_MAKER',
        'MARKET',
        'STOP_LOSS_LIMIT',
        'TAKE_PROFIT_LIMIT'
      ],
      icebergAllowed: true,
      ocoAllowed: true,
      quoteOrderQtyMarketAllowed: true,
      isSpotTradingAllowed: true,
      isMarginTradingAllowed: true,
      filters: [
        {
          filterType: 'PRICE_FILTER',
          minPrice: '0.01000000',
          maxPrice: '1000000.00000000',
          tickSize: '0.01000000'
        },
        {
          filterType: 'PERCENT_PRICE',
          multiplierUp: '5',
          multiplierDown: '0.2',
          avgPriceMins: '5'
        },
        {
          filterType: 'LOT_SIZE',
          minQty: '0.00001000',
          maxQty: '9000.00000000',
          stepSize: '0.00001000'
        },
        {
          filterType: 'MIN_NOTIONAL',
          minNotional: '10.00000000',
          applyToMarket: true,
          avgPriceMins: '5'
        },
        { filterType: 'ICEBERG_PARTS', limit: '10' },
        {
          filterType: 'MARKET_LOT_SIZE',
          minQty: '0.00000000',
          maxQty: '186.34181397',
          stepSize: '0.00000000'
        },
        { filterType: 'MAX_NUM_ORDERS', maxNumOrders: '200' },
        { filterType: 'MAX_NUM_ALGO_ORDERS', maxNumAlgoOrders: '5' }
      ],
      permissions: [ 'SPOT', 'MARGIN' ]
    },
    limits: {
      price: { min: 0.01, max: 1000000 },
      amount: { min: 0.00001, max: 9000 },
      market: { min: 0, max: 186.34181397 },
      cost: { min: 10 }
    }
  }
}
```

**Pull request closed without merging**